components/plurals/src/rules/runtime/ast.rs

```diff
@@ -0,0 +1,634 @@
+            AndOr::And
+        } else {
+            AndOr::Or
+        };
+
+        let polarity = if encoded & 0b0100_0000 != 0 {
+            Polarity::Positive
+        } else {
+            Polarity::Negative
+        };
+
+        let operand = Operand::from_idx(encoded & 0b0011_1111);
+        (and_or, polarity, operand)
+    }
+}
+
+unsafe impl VarULE for RelationULE {
+    type Error = &'static str;
+
+    #[inline]
+    unsafe fn from_byte_slice_unchecked(bytes: &[u8]) -> &Self {
+        let ptr = bytes.as_ptr();
+        let len = bytes.len();
+        // subtract length of andor_polarity_operand and modulo and then convert between a slice of bytes and PlainOldULE<8>
+        let len_new = (len - 5) / 8;
+        // it's hard constructing custom DSTs, we fake a pointer/length construction
+        // eventually we can use the Pointer::Metadata APIs when they stabilize
+        let fake_slice = core::ptr::slice_from_raw_parts(ptr as *const RangeOrValueULE, len_new);
+        let ret = &*(fake_slice as *const Self);
+        debug_assert_eq!(core::mem::size_of_val(ret), core::mem::size_of_val(bytes));
+        ret
+    }
+
+    #[inline]
+    fn as_byte_slice(&self) -> &[u8] {
+        unsafe {
+            core::slice::from_raw_parts(
+                self as *const Self as *const u8,
+                core::mem::size_of_val(self),
+            )
+        }
+    }
+
+    //     what it should do is attempt to parse the first 4 bytes as a u32::ULE (POU<4>), and the remaining bytes as a ZV<RangeOrValueULE>
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let bits = bytes[0]; // XXX: Validate those bits
+        <PlainOldULE<4> as ULE>::validate_byte_slice(&bytes[1..5]).map_err(|_| "foo")?;
+        let remaining = &bytes[5..];
+        RangeOrValueULE::validate_byte_slice(remaining).map_err(|_| "foo")?;
+        Ok(())
+    }
+}
+
+unsafe impl EncodeAsVarULE<RelationULE> for Relation<'_> {
+    fn encode_var_ule_as_slices<R>(&self, cb: impl FnOnce(&[&[u8]]) -> R) -> R {
+        let encoded =
+            RelationULE::encode_andor_polarity_operand(self.and_or, self.polarity, self.operand);
+        cb(&[
+            &[encoded],
+            PlainOldULE::<4>::as_byte_slice(&[self.modulo.as_unaligned()]),
+            self.range_list.as_bytes(),
+        ])
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct RangeOrValueULE(PlainOldULE<8>);
+
+unsafe impl ULE for RangeOrValueULE {
+    type Error = zerovec::ule::ULEError<core::convert::Infallible>;
+
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        PlainOldULE::<8>::validate_byte_slice(bytes)
+    }
+}
+
+impl AsULE for RangeOrValue {
+    type ULE = RangeOrValueULE;
+
+    #[inline]
+    fn as_unaligned(self) -> Self::ULE {
+        let mut result = [0; 8];
+        match self {
+            Self::Range(start, end) => {
+                let start_bytes = start.to_be_bytes();
```

Comment:
Nit, here and below: it's weird, although not wrong, that you are using `to_be_bytes()` in a ULE impl.  I'd just use `to_le_bytes()`.  You might even get a small performance boost on LE platforms if you do that.

Reply:
I followed what I did in `datetime`. It would be nice to settle on the single model, and IIRC `to_be_bytes` was the preferred one when we discussed it in context of datetime ULE.

Reply:
I see the datetime ULE as a special case because we had a unique constraint, which is that the top bit of the first byte had to be unassigned so we could use it for the discriminant, which only BE guaranteed. But for an entire 4-byte number with no special cases, I think we should generally use LE.